File: python/species.py
import re

class Element:
    def __init__(self, name="", mass=0.0, charge=0.0, atomic_number=0):
        self.name = name
        self.mass = mass
        self.charge = charge
        self.atomic_number = atomic_number

class Species:
    def __init__(self):
        self.number_of_elements = 0
        self.ele_data = []

    def __del__(self):
        pass

    def load_species(self, in_stream, num):
        self.number_of_elements = num

        for _ in range(self.number_of_elements):
            line = in_stream.readline().strip()
            words = self.split(line)

            ele = Element()
            ele.name = words[0]
            ele.mass = float(words[1])
            ele.charge = float(words[2])
            if len(words) > 3:
                ele.atomic_number = int(words[3])

            self.ele_data.append(ele)

    def read_species(self, in_stream):

        for _ in range(10000):
            line = in_stream.readline().strip()
            words = self.split(line)
            if words[0].lower() == "species":
                self.number_of_elements = int(words[1])
                break

        for _ in range(self.number_of_elements):
            line = in_stream.readline().strip()
            words = self.split(line)

            ele = Element()
            ele.name = words[0]
            ele.mass = float(words[1])
            ele.charge = float(words[2])
            if len(words) > 3:
                ele.atomic_number = int(words[3])

            self.ele_data.append(ele)

    def print_species(self, out_stream):
        out_stream.write("\n \n *********************************************************************\n")
        out_stream.write("                         atom data \n")
        out_stream.write("\n \n *********************************************************************\n")
        out_stream.write("\n        element         charge           mass\n")

        for ele in self.ele_data:
            out_stream.write(f"{ele.name:15} {ele.charge:15.3f} {ele.mass:15.3f}\n")

    def get_species(self, i) -> Element:
        return self.ele_data[i]
    
    def get_num_species(self):
        return self.number_of_elements

    def get_mass(self, i):
        return self.ele_data[i].mass

    def split(self, s):
        return re.split(r'\s+', s.strip())

# Example usage:
# To load data from a file, you can use the following:
# with open('data.txt', 'r') as infile:
#     species = Species()
#     species.load_species(infile, num_elements)
# To print data to a file, you can use:
# with open('output.txt', 'w') as outfile:
#     species.print_species(outfile)

File: potentials/metal_manybody/src/Species.cpp
// species.cpp: implementation of the Species class.
//
//////////////////////////////////////////////////////////////////////

#include "Species.h"

std::vector<std::string> split(std::string s);

//////////////////////////////////////////////////////////////////////
// construction/destruction
//////////////////////////////////////////////////////////////////////
/* default constructor */
Species::Species()
{
    numberOfElements = 0;
}

/* default destructor */
Species::~Species()
{

}

/** dimensions and loads elemental data from file
 * m_element, m_mass, m_charge assumed to be on the same line
 * receives the number of different elements present and
 * the file to read from.
 * returns integer for error checking
 */

void Species::loadSpecies(std::ifstream& inStream, int num)
{
    int
        i;

    std::string
        dummy,
        line;

    Element
        ele;

    std::vector<std::string> words;

    numberOfElements = num;

    for (i = 0; i < numberOfElements; i++)
    {

        std::getline(inStream, line);
        //std::cout << "species line: " << line << std::endl;
        words = split(line);

        ele.name = words[0];

        dummy = words[1];
        ele.mass = std::stod(dummy);

        dummy = words[2];
        ele.charge = std::stod(dummy);

	    if (words.size() > 3)
        {
            dummy = words[3];
            ele.atomicNumber = std::stoi(dummy);
        }

        eleData.push_back(ele);
        // error checking

    }

}


/**
  function to print species data
 **/

void Species::printSpecies(std::ofstream& outstream)
{
    Element
        ele;

    outstream << "\n \n *********************************************************************" << std::endl;
    outstream << "                         atom data ";
    outstream << "\n \n *********************************************************************" << std::endl;

    outstream << "\n        element         charge           mass" << std::endl;

    for (int i = 0; i < numberOfElements; i++)
    {
        ele = eleData[i];

        outstream << std::setiosflags(std::ios::showpoint | std::ios::fixed)
            << std::setprecision(3) << "\n"
            << std::setw(15) << ele.name
            << std::setw(15) << ele.charge
            << std::setw(15) << ele.mass << std::endl;
    }

    outstream.flush();

}


/************************************************************
 returns the data for element i
 ***********************************************************/
Element Species::getSpecies(int i) const
{
    Element
        dummy;

    dummy = eleData[i];
    return dummy;
}



/************************************************************
 returns the mass of an element
 ***********************************************************/
double Species::getMass(int i)
{
    return eleData[i].mass;
}
